### ISSUE TITLE
Добавить скачивание в многопоточном режиме

### ISSUE THREAD

**Issue by carloscondit** [Solbreen/Kaznacheistvo#1]
Исходя из API, после скачивания первой страницы с данными мы узнаем сколько всего страниц и сколько всего записей содержится в этих страницах. Соответственно можно разделить страницы на группы и создать фоновую задачу для скачивания страниц из каждой группы. А после завершения всех фоновых задач собрать их результаты и обработать. Это ускорить скачивание в несколько раз. Общее число записей на ресурсе может пригодится в конце скрипта, чтобы убедится в том, что все страницы были скачаны и все записи обработаны.

**Comment by Solbreen** [comment]
Ваш запрос обработан. Было выполнено обновление файла KazFastmm.ps1 . На данный момент добавлено многопоточное скачивание

**Comment by carloscondit** [comment]
А получится использовать общее число записей на ресурсе для сверки с обработанными записями? Если они будут отличаться, то можно будет предположить, что в процессе обработки или скачивания были ошибки.

**Comment by Solbreen** [comment]
> А получится использовать общее число записей на ресурсе для сверки с обработанными записями? Если они будут отличаться, то можно будет предположить, что в процессе обработки или скачивания были ошибки.

Есть у меня идея, но я не совсем уверен, что это то, что нужно. Можно сделать счетчик, который считает каждую запись. В конце сравниваем насчитанное с числом записей на ресурсе.
Это то, что нужно?

**Comment by carloscondit** [comment]
> > А получится использовать общее число записей на ресурсе для сверки с обработанными записями? Если они будут отличаться, то можно будет предположить, что в процессе обработки или скачивания были ошибки.
> 
> Есть у меня идея, но я не совсем уверен, что это то, что нужно. Можно сделать счетчик, который считает каждую запись. В конце сравниваем насчитанное с числом записей на ресурсе. Это то, что нужно?

Это лишнее действие. Массив data уже имеет нужное тебе свойство.

**Comment by Solbreen** [comment]
Готово

**Comment by carloscondit** [comment]
Верно. Но добавь в этот же блок finaly сообщение об этом действии в поток Verbose. Соответственно в этом сообщении должно быть видно какой каталог удаляется. Когда будешь закрывать другие запросы, то попробуй найти способ вставлять ссылки на эти изменения.

**Comment by carloscondit** [comment]
Все верно. Но предлагаю исправить текст на: обработанное количество записей не совпадает с количеством записей на ресурсе. А так же добавь сообщение в поток Warning, в котором отрази количество обработанных записей и количество записей на ресурсе казначейства. Можно как вариант throw заменить на Write-Error.

**Comment by Solbreen** [comment]
Почему именно в Warning?
Я так понимаю, что несовпадение записей должно быть ошибкой, тк при несовпадении данные будут получены ошибочные. Я думаю, что в таком случае программа должна быть остановлена, ну или  по крайней мере можно предложить пользователю выбрать продолжать или нет. Но скрипт то задумывался выполняться без участия пользователя.
Важно ли, чтобы при возникновении ситуации несовпадения где то сохранялось сообщение о возникновении такой ситуации?

**Comment by carloscondit** [comment]
Поток warning по умолчанию выводится в консоль, в отличии от Verbosе. Если бы у нас было настроено логирование работы скрипта (весь вывод скрипта в текст), то это сообщение попало бы в лог. Throw здесь правильно стоит, работа скрипта после этого завершиться. Write-Warning будет строкой выше и просто будет информировать о проблеме и в частности что и с чем не совпадает. Подразумевается что пользователь настроит работу скрипта по расписанию, но другим скриптом. То есть твой скрипт править никто не будет.

1. Управляюющий скрипт вызывает твой скрипт с нужными параметрами
2. Если твой скрипт завершается с ошибкой, то управляющий скрипт запишет текст с консоли (а там будет сообщение Warning) в лог файл и отправить уведомление на почту
3. Пользователь проверит лог и будет искать причину ошибки

**Comment by Solbreen** [comment]
готово, вот [коммит](https://github.com/Solbreen/Kaznacheistvo/commit/a192182b5f1d2780b247078f3079fbfeccc5fd55)
